core/src/pretty.rs

```diff
@@ -152,17 +153,197 @@ fn needs_parens_in_type_pos(typ: &Type) -> bool {
     }
 }
 
-pub fn fmt_pretty<'a, T>(value: &T, f: &mut fmt::Formatter) -> fmt::Result
+pub fn fmt_pretty<T>(value: &T, f: &mut fmt::Formatter) -> fmt::Result
 where
-    T: Pretty<'a, pretty::BoxAllocator, ()> + Clone,
+    T: for<'a> Pretty<'a, BoundedAllocator, ()> + Clone,
 {
-    let doc: DocBuilder<_, ()> = value.clone().pretty(&pretty::BoxAllocator);
+    let allocator = BoundedAllocator::default();
+    let doc: DocBuilder<_, ()> = value.clone().pretty(&allocator);
     doc.render_fmt(80, f)
 }
 
-impl<'a, A: Clone + 'a> NickelAllocatorExt<'a, A> for pretty::BoxAllocator {}
+#[derive(Clone, Copy, Debug, Default)]
+struct SizeBound {
+    depth: usize,
+    size: usize,
+}
+
+/// A guard for scoping changes to a `SizeBound`, restoring the old
+/// value when leaving a scope.
+#[derive(Default)]
+struct ShrinkGuard<'a> {
+    bound: Option<&'a Cell<SizeBound>>,
+    old_bound: SizeBound,
+}
+
+impl<'a> Drop for ShrinkGuard<'a> {
+    fn drop(&mut self) {
+        if let Some(bound) = self.bound {
+            bound.set(self.old_bound);
+        }
+    }
+}
+
+/// A pretty-printing allocator that supports rough bounds on the
+/// size of the output.
+///
+/// When a pretty-printed object is too large, it will be abbreviated.
+/// For example, a record will be abbreviated as "{…}".
+///
+/// The bounds are "rough" in that the depth bound only (currently; this might
+/// be extended in the future) constrains the number of nested records: you can
+/// still have deeply nested terms of other kinds. The size bound only constrains
+/// the number of children of nested records. As such, neither constraint gives
+/// precise control over the size of the output.
+pub struct BoundedAllocator {
+    inner: pretty::BoxAllocator,
+    bound: Option<Cell<SizeBound>>,
+}
+
+/// The default `BoundedAllocator` imposes no constraints.
+impl Default for BoundedAllocator {
+    fn default() -> Self {
+        Self {
+            inner: pretty::BoxAllocator,
+            bound: None,
+        }
+    }
+}
+
+impl BoundedAllocator {
+    /// Creates a `BoundedAllocator` with constraints.
+    pub fn bounded(max_depth: usize, max_size: usize) -> Self {
+        Self {
+            inner: pretty::BoxAllocator,
+            bound: Some(Cell::new(SizeBound {
+                depth: max_depth,
+                size: max_size,
+            })),
+        }
+    }
+
+    /// Shrinks this allocator, returning a guard that restores the
+    /// old constraints when it leaves scope.
+    fn shrink(&self, child_size: usize) -> ShrinkGuard<'_> {
```

Comment:
I think the RAII pattern is cool, but I'm concerned that the `shrink_guard` is never used by anything. Can we thus guarantee that the compiler won't drop it before the end of the scope? Or even will just get rid of it entirely and never create the object as technically dead code?

I'm not really sure how to look at that beside looking at the generated ASM (but even that isn't a guarantee). Another approach could be to do the same as for a mutex: we could borrow `self`, even mutably - to make sure we don't use the original while a shrinked version is used - and then expose `self` as an immutable reference inside `ShrinkGuard`, so all usages of `shrinked` self should go through the guard. Exactly as a mutex guard, in the end.

Reply:
Good to know that there are some guarantees. However, I still find it really strange to create a guard that we never use and somehow to still rely on its scope, as a user of that API.

If we just want scoping, what about spawning a closure instead?

Like

```rust
allocator.shrinked(|| {
  do_stuff with shrinked self
)
```

We can pass `&self` as an argument as well, or not, it doesn't really matter. At least it makes it clear where the scope is delimited and that using any immutable reference to `self` doesn't make any difference (whether it comes from the external or not).

Reply:
Ah, the closure API seems nicer! I'll give that a try